cranelift/isle/isle/src/trie_again.rs

```diff
@@ -0,0 +1,543 @@
+    /// `Option::Some` and produce a new binding from its contents. (This currently only happens
+    /// with external extractors.)
+    Some {
+        /// Which binding is being matched?
+        source: BindingId,
+    },
+    /// After some sequence of matches, we'll match one of the previous bindings against a tuple and
+    /// produce a new binding from one of its fields. (This currently only happens with external
+    /// extractors.)
+    Tuple {
+        /// Which binding is being matched?
+        source: BindingId,
+        /// Which tuple field are we projecting out?
+        field: TupleIndex,
+    },
+}
+
+/// Pattern matches which can fail.
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub enum Constraint {
+    /// The value must match this enum variant.
+    Variant(TupleIndex, sema::TypeId, sema::VariantId),
+    /// The value must equal this integer literal.
+    ConstInt(i128),
+    /// The value must equal this Rust primitive value.
+    ConstPrim(sema::Sym),
+    /// The value must be an `Option::Some`, from a fallible extractor.
+    Some,
+}
+
+/// Expressions construct new values. Rust pattern matching can only destructure existing values,
+/// not call functions or construct new values. So `if-let` and external extractor invocations need
+/// to interrupt pattern matching in order to evaluate a suitable expression. These expressions are
+/// also used when evaluating the right-hand side of a rule.
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub enum Expr {
+    /// Evaluates to the given integer literal.
+    ConstInt(i128),
+    /// Evaluates to the given primitive Rust value.
+    ConstPrim(sema::Sym),
+    /// One of the arguments to the top-level function.
+    Argument(TupleIndex),
+    /// A binding from some sequence of pattern matches.
+    Binding(BindingId),
+    /// The result of calling an external extractor.
+    Extractor(sema::TermId, ExprId),
+    /// The result of constructing an enum variant.
+    Variant(sema::TypeId, sema::VariantId, Box<[ExprId]>),
+    /// The result of calling an external constructor.
+    Constructor(sema::TermId, Box<[ExprId]>),
+}
+
+/// A term-rewriting rule. All [BindingId]s and [ExprId]s are only meaningful in the context of the
+/// [RuleSet] that contains this rule.
+#[derive(Debug, Default)]
+pub struct Rule {
+    /// Where was this rule defined?
+    pub pos: Pos,
+    /// All of these bindings must match for this rule to apply. Note that within a single rule, if
+    /// a binding site must match two different constants, then the rule can never match.
+    pub constraints: HashMap<BindingId, Constraint>,
+    /// If (a, b) is in this map, then binding sites `a` and `b` must be equal for this rule to
+    /// match. As an additional invariant, these pairs must form a circularly-linked list for each
+    /// equivalence class, so we can find all the binding sites in that class given any one of them.
+    pub equals: HashMap<BindingId, BindingId>,
```

Comment:
Yeah, I thought about using union-find until I realized I needed to visit every element of the equivalence class starting from an arbitrary element of the class. Although I'm not 100% sure that's as important as I thought it was, so union-find might still be an option.

I'm amused to realize that union-find plus a set for each root is _also_ O(1) space per binding, it just has larger constants... I think you could maintain first-child/next-sibling pointers in a union-find data structure, which would only triple the constant factors over my version.

I'm not immediately convinced it's worth factoring out a separate type for this though. I'm inclined to see if it gets more complex in the rest of review first.

Reply:
I went ahead and used union-find, factored out to a separate type since that's a little more complex.

It's a good change, except that now removing a set from the data structure requires visiting every member of every set to figure out whether it's part of the desired set.

I'm inclined to not try to improve that, but I have some ideas if we find it's a performance problem later.